burn-ndarray/src/ops/unfold.rs

```diff
@@ -0,0 +1,91 @@
+/// # Notes
+/// The idea behind using convolution for unfolding is to leverage the sliding window mechanism of
+/// convolution. By creating a weight tensor with ones in a particular pattern, we are able to borrow
+/// the convolution operation's mechanism as it moves across the input tensor, picking up the desired
+/// values in the pattern of the unfolding operation.
+fn create_unfolding_weight<E: FloatNdArrayElement>(
+    in_channels: usize,
+    kernel_size: [usize; 2],
+) -> NdArrayTensor<E, 4> {
+    let mut weight = Array4::zeros(Dim([
+        in_channels * kernel_size[0] * kernel_size[1],
+        in_channels,
+        kernel_size[0],
+        kernel_size[1],
+    ]));
+
+    for k in 0..in_channels {
+        for i in 0..kernel_size[0] {
+            for j in 0..kernel_size[1] {
+                let output_channel = k * kernel_size[0] * kernel_size[1] + i * kernel_size[1] + j;
+                weight[[output_channel, k, i, j]] = E::one();
+            }
+        }
+    }
+
+    NdArrayTensor::new(weight.into_dyn().into_shared())
+}
+
+pub(crate) fn unfold4d<E: FloatNdArrayElement>(
+    x: NdArrayTensor<E, 4>,
+    kernel_size: [usize; 2],
+    options: UnfoldOptions,
+) -> NdArrayTensor<E, 3> {
+    let [batch_size, in_channels, in_height, in_width] = x.shape().dims;
+    let stride = options.stride.unwrap_or([1, 1]);
+    let padding = options.padding.unwrap_or([0, 0]);
+    let dilation = options.dilation.unwrap_or([1, 1]);
+
+    // See https://pytorch.org/docs/stable/generated/torch.nn.Unfold.html#torch.nn.Unfold for full explanation,
+    // This calculates the number of patches with each patch having channels_out values
+    let l_dim_1 =
+        (in_height + 2 * padding[0] - dilation[0] * (kernel_size[0] - 1) - 1) / stride[0] + 1;
+    let l_dim_2 =
+        (in_width + 2 * padding[1] - dilation[1] * (kernel_size[1] - 1) - 1) / stride[1] + 1;
+    let l = l_dim_1 * l_dim_2;
+    let channels_out = in_channels * kernel_size[0] * kernel_size[1];
+
+    let weight = create_unfolding_weight(in_channels, kernel_size);
+    let unfolded = conv2d(
+        x,
+        weight,
+        None,
+        ConvOptions {
+            stride,
+            padding,
+            dilation,
+            groups: 1,
+        },
+    );
+
+    let inner_array = unfolded.array;
+    let mut reshaped_array = Array3::zeros([batch_size, channels_out, l]);
+
+    // Iterate over each dimension and fill in the values from unfolded to reshaped
+    for b in 0..batch_size {
+        for c in 0..channels_out {
+            let mut l_index = 0; // Index to keep track of where to put the value in the L dimension
+            for h in 0..l_dim_1 {
+                for w in 0..l_dim_2 {
+                    // Copy the value from unfolded to reshaped
+                    let value = inner_array[[b, c, h, w]];
+                    reshaped_array[[b, c, l_index]] = value;
+
+                    l_index += 1;
+                }
+            }
+        }
+    }
```

Comment:
Does a reshape does the same thing? `unfolded.reshape([batch_size, channels, dim1 * dim2])` Maybe we need to swap the two axes before, but it's probably way more efficient to do it that way.

Reply:
@nathanielsimard Unfortunately no. The short explanation is that the number of elements before and after a `reshape` needs to be the same, but an `unfold` doesn't do that. If we look at PyTorch's example, a `[2, 5, 3, 4]`(120 elements) tensor unfolded with a kernel size of `[2, 3]` and default options leads to an unfolded tensor with dimensions `[2, 30, 4]` (240 elements).

This [Stack Overflow thread](https://stackoverflow.com/questions/53972159/how-does-pytorchs-fold-and-unfold-work) can probably give a better intuition as to why. This nice graphic from the top answer really nails why (although note that this is for _tensor unfolds_, not _nn unfolds_, which is what we're doing here. A nn unfold extracts spatial patches not "patches" along a dimension):

![image](https://github.com/burn-rs/burn/assets/55096933/0d3dcbe2-69ac-4256-aa12-040a6ab2d4a2)

That being said, I agree that this way of extracting spatial patches is not particularly efficient, so I'm very open to other ways of doing this.

Reply:
I think the size of the convolution matches the size of the output from the unfold operation:

From Pytorch:
Conv2d:

![image](https://github.com/burn-rs/burn/assets/14095719/d55ec737-8a64-43d7-93a5-b298534fccf9)

Unfold:
![image](https://github.com/burn-rs/burn/assets/14095719/379ce3a4-9a7f-4b17-944a-c28c91bc7d80)


I tested using a simple reshape at the end (it's just a flatten) and the tests were passing, so I think it should always work!

Reply:
What I tested with:

```rust
pub(crate) fn unfold4d<E: FloatNdArrayElement>(
    x: NdArrayTensor<E, 4>,
    kernel_size: [usize; 2],
    options: UnfoldOptions,
) -> NdArrayTensor<E, 3> {
    let [_batch_size, in_channels, _in_height, _in_width] = x.shape().dims;
    let stride = options.stride.unwrap_or([1, 1]);
    let padding = options.padding.unwrap_or([0, 0]);
    let dilation = options.dilation.unwrap_or([1, 1]);

    let weight = create_unfolding_weight(in_channels, kernel_size);
    let unfolded = conv2d(
        x,
        weight,
        None,
        ConvOptions {
            stride,
            padding,
            dilation,
            groups: 1,
        },
    );

    let [batch_size, channels_out, out_height, out_width] = unfolded.shape().dims;

    NdArrayOps::reshape(
        unfolded,
        Shape::new([batch_size, channels_out, out_height * out_width]),
    )
}
```

Reply:
Yes you're right, I'll switch to using a `reshape` instead!